### PR TITLE
LibTx: Support multi-codepoint placeholders

### PR DESCRIPTION
**Pull request by GaryQian**
Adds text layout level support for handling placeholders that represent multiple codepoints.

Needed for https://github.com/flutter/flutter/pull/84986 and clipboard behavior to function properly.

Previously, we represented each placeholder with a single object replacement character 0xFFFC. This patch allows each placeholder to specify how many replacement characters to use to represent it. This allows operations such as GetRectsForRange or GetPositionForOffset to produce and handle indexes that work cleanly with higher level concepts such as TextEditingValue which must use multi-char strings to represent placeholders.

This needs to be replicated in SkParagraph.

### CODE REVIEW

**Comment by GaryQian** [comment]
@Rusino FYI

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 1]
```diff
@@ -968,26 +971,26 @@ void ParagraphTxt::Layout(double width) {
             // The placeholder run's layout should yield one glyph representing
             // the object replacement character.  Replace its width with the
             // placeholder's width.
-            FML_DCHECK(layout.nGlyphs() == 1);
             glyph_advance = run.placeholder_run()->width;
+            glyph_positions.emplace_back(
```
Comment (jason-simmons) [inline third_party/txt/src/txt/paragraph_txt.cc:975]: Would it be possible to insert the placeholder's `codepoint_length` into `grapheme_code_unit_widths` and then use the current logic for populating `glyph_positions`?

Comment (GaryQian) [reply third_party/txt/src/txt/paragraph_txt.cc:975]: I think the goal is to have all of the obj replacement chars to be "stacked" on top of each other with the same coordinates. This way, indexing any of the chars in the range will result in the same results.

**Comment by GaryQian** [comment]
cc @mdebbar for web side. Does this look doable in the web implementation?

File: third_party/txt/src/txt/paragraph_txt.cc (hunk 1 above)
Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph_txt.cc:975]: What is the difference between that and a single glyph that consumes `codepoint_length` characters?

**Comment by mdebbar** [comment]
@GaryQian currently, placeholders are handled kind of like a zero-length span. I'm sure it's possible to support variable length placeholders if we implement it carefully.

We need to do this for the CanvasKit renderer too, which I'm not very familiar with. cc @hterkelsen.

**Comment by GaryQian** [comment]
@Renzo-Olivares Added tests to verify zero-size multi-codepoint placeholders work properly :)

**Comment by zanderso** [comment]
This PR appears to be obsolete.

**Pull request closed without merging**